### ISSUE TITLE
VMware: Cross-vCenter vm/template clone

### ISSUE THREAD

**Issue by balonik** [ansible/ansible#53693]
##### SUMMARY

It would be great to have module or functionality added to some existing module that could clone vm or template to a linked (or other) vCenter.
Googling around this should be available in pyvmomi already: https://github.com/vmware/pyvmomi/issues/490

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

vmware_guest
or new module

##### ADDITIONAL INFORMATION

```yaml

```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/vmware/vmware_guest.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @Tomorrow9 @ckotte @dav1x @lparkes @nerzhul @pdellaert @warthog9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@balonik, just so you are aware we have a dedicated Working Group for vmware.
You can find other people interested in this in `#ansible-vmware` on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @goneri
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @pgbidkar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Akasurde** [comment]
resolved_by_pr #66156

**Comment by Akasurde** [comment]
@balonik, Could you please check PR #66156 and let us know if it works for you ? <br><br>Please check [How to test a PR](https://docs.ansible.com/ansible/latest/dev_guide/testing.html#how-to-test-a-pr) for more help. <br><br> Thanks. <br><br> needs_info

**Comment by balonik** [comment]
@Akasurde 

During my testing I have find out following cosmetic issues:

- if you specify non-existent `destination_vm_folder` you get rather ugly error from pyvmomi. Could help if  `find_folder_by_name` returned value is checked and module exits properly. Snippet from the error below.
```
File \"/tmp/ansible_vmware_guest_cross_vc_clone_payload_1QGV2O/ansible_vmware_guest_cross_vc_clone_payload.zip/ansible/modules/cloud/vmware/vmware_guest_cross_vc_clone.py\", line 233, in clone\n  File \"/usr/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py\", line 706, in <lambda>\n    self.f(*(self.args + (obj,) + args), **kwargs)\n  File \"/usr/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py\", line 511, in _InvokeMethod\n    list(map(CheckField, info.params, args))\n  File \"/usr/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py\", line 1041, in CheckField\n    raise TypeError('Required field \"%s\" not provided (not @optional)' % info.name)\nTypeError: Required field \"folder\" not provided (not @optional)\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1
```
- it would help if I could specify ESXi Cluster instead of ESXi host

I am using specific role for the user and I am currently struggling to find out which permission I am missing since I am getting 'Permission to perform this operation was denied.' If you have this info somewhere could you post it here? The module works using admin user.

**Comment by Akasurde** [comment]
@anusha94 Could you please take a look into these concerns? Thanks in advance.

**Comment by anusha94** [comment]
@balonik ,
Thanks for the feedback, I will look into those issues.

**Comment by balonik** [comment]
@anusha94 

- If VM with the name already exists module fails with `FAILED! => {"changed": false, "msg": "A VM with the given name already exists"}` instead of `no action is taken` written in documentation.

**Comment by anusha94** [comment]
@balonik Can you share what is the permission level for your user roles?
I believe the permission level should be the same as you would have for deploying a vm or for vmotion

@Akasurde is there any specific way to document this in Ansible? I don't see other modules listing out roles

**Comment by Akasurde** [comment]
@anusha94 You can document this in module documentation itself like [`vmware_guest`](https://github.com/ansible/ansible/blob/3bf8b1d1c99cf5354cb7687d4a8669a144f3f90d/lib/ansible/modules/cloud/vmware/vmware_guest.py#L33) or in [scenarios guides](https://docs.ansible.com/ansible/latest/scenario_guides/vmware_scenarios/scenario_clone_template.html#scenario-requirements). Thanks.

**Comment by balonik** [comment]
@anusha94 
I was able to narrow my permission issues to DVSwitch configuration on the source VM.
If the source VM has DVSwitch assigned to NIC and such DVSwitch is not on the destination vCenter (I guess that is the case all the time) vCenter gives you following warning in the UI:

```
Network interface 'Network adapter 1' uses network 'DVSwitch: 1e b5 1e 50 5d c1 1a d1-6e 32 7f 13 8b d9 33 1a', which is not accessible.
```

vCenter seems to workaround this warning by removing NIC assignment from source VM before the clone proces starts. After the VM Clone is done, the cloned VM (on destination) has 'Network adapter 1' assignment empty.

If I set the 'Network adapter 1' to 'VM Network' before VM clone, I don't get permission denied and the cloned VM has the same network config.
